dpnp/backend/extensions/indexing/choose.cpp

```diff
@@ -0,0 +1,453 @@
+//*****************************************************************************
+// Copyright (c) 2024, Intel Corporation
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+// - Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+// - Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <sycl/sycl.hpp>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "choose_kernel.hpp"
+#include "dpctl4pybind11.hpp"
+#include "utils/indexing_utils.hpp"
+#include "utils/memory_overlap.hpp"
+#include "utils/output_validation.hpp"
+#include "utils/sycl_alloc_utils.hpp"
+#include "utils/type_dispatch.hpp"
+
+namespace dpnp::extensions::indexing
+{
+
+namespace td_ns = dpctl::tensor::type_dispatch;
+
+static kernels::choose_fn_ptr_t choose_clip_dispatch_table[td_ns::num_types]
+                                                          [td_ns::num_types];
+static kernels::choose_fn_ptr_t choose_wrap_dispatch_table[td_ns::num_types]
+                                                          [td_ns::num_types];
+
+namespace py = pybind11;
+
+/*
+ Returns an std::unique_ptr wrapping a USM allocation and deleter.
+
+ Must still be manually freed by host_task when allocation is needed
+ for duration of asynchronous kernel execution.
+*/
+template <typename T>
+auto usm_unique_ptr(std::size_t sz, sycl::queue &q)
+{
+    using dpctl::tensor::alloc_utils::sycl_free_noexcept;
+    auto deleter = [&q](T *usm) { sycl_free_noexcept(usm, q); };
```

Comment:
Please be very careful with capturing parameters by reference. You always must take into consideration theirs lifetime.
It is better to capture `q` as copy in this particular case.
`sycl::queue` is a smart pointer anyway.  It is cheap to make a copy.